### PR TITLE
Generate SSNs in post-processing

### PR DESCRIPTION
**Pull request by mattkappel**
### Description
- *Category*:  implementation
- *JIRA issue*: [MIC-3644](https://jira.ihme.washington.edu/browse/MIC-3644) and [MIC-3837](https://jira.ihme.washington.edu/browse/MIC-3837)
- *Research reference*: https://vivarium-research.readthedocs.io/en/latest/models/concept_models/vivarium_census_synthdata/concept_model.html#social-security-number-ssn-and-individual-taxpayer-identification-number-itin

### Changes and notes
- Removes old SSN implementation and tests
- Renames boolean `ssn` to `has_ssn` in observers
- Generates unique SSNs for those simulates who have one or an empty string if no SSN
- Adds a `simulant_id` mapper layer to be later used to map ITINs
- Adds tests for SSN uniqueness, correctness, and ranges
- Add retry mechanism to ITIN generation as apparent standard
- Changes ITIN uniqueness test to allow no duplicates (previously allowed 5)

### Verification and Testing
CI, newly added and previously extant, succeeds.

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 1]
```diff
@@ -1,19 +1,62 @@
...
 from vivarium.framework.randomness import RandomnessStream
 
+from vivarium_census_prl_synth_pop.results_processing.formatter import (
+    format_data_for_mapping,
+)
 from vivarium_census_prl_synth_pop.utilities import random_integers
 
 
-def generate_itin(
+def get_simulant_id_maps(
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:15]: Consider renaming to soemthing like `get_simulant_id_ssn_maps`

Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:15]: No, this is meant as a top-level for anything that is indexed by `simulant_id`. Eventually we'll add ITIN.

Comment (albrja) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:15]: @stevebachmeier The way to think about this is in make_results we are mapping the `simulant_id `column.  If we need to map that column to multiple columns for results (in this case SSNs and ITINs) we need a main runner to get each of the maps from `simulant_id`.

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 2]
```diff
@@ -1,19 +1,62 @@
...
+        raise ValueError(f"Expected `simulant_id`, got `{column_name}`")
+    maps = dict()
+    output_cols_superset = [column_name, "has_ssn"]
+    formatted_obs_data = format_data_for_mapping(
+        index_name=column_name,
+        obs_results=obs_data,
+        output_columns=output_cols_superset,
+    )
+    maps.update(get_ssn_map(column_name, formatted_obs_data, randomness))
+    return maps
```
Comment (albrja) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:50]: why are you not mapping ITINs here?

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 3]
```diff
@@ -24,46 +67,166 @@ def generate_itin(
...
-        randomness=randomness,
-        additional_key=f"{additional_key}_itin_serial",
+        area = random_integers(
+            min_val=1,
+            max_val=899,
+            index=ssn.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_area",
+        )
+        area = np.where(area == 666, 667, area)  # XXX is this reasonable?
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:169]: why no 666?

Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:169]: Oh, nm - they just don't assign that prefix. Yeah, seems reasonable unless `area.replace(666, 667)` is more performant

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 4]
```diff
@@ -24,46 +67,166 @@ def generate_itin(
...
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_serial",
+        )
+        ssn["ssn_serial"] = serial
+
+        ssn["ssn"] = ssn.ssn_area.astype(str).str.zfill(3) + "-"
+        ssn["ssn"] += ssn.ssn_group.astype(str).str.zfill(2) + "-"
+        ssn["ssn"] += ssn.ssn_serial.astype(str).str.zfill(4)
+
+        return ssn["ssn"].to_numpy()
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:194]: Might just do `return (ssn_area.map(str) + "-" + ssn_group.map(str) + "-" + ssn_series.map(str)).to_numpy()`

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 5]
```diff
@@ -1,19 +1,62 @@
...
+    return maps
+
+
+def generate_itins(
     size: int,
     additional_key: Any,
     randomness: RandomnessStream,
 ) -> np.ndarray:
     """
-    Generate a np.ndarray of length `size` of hopefully unique ITIN values.
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:16]: lol

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py (hunk 2 above)
Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:50]: My understanding for MIC-3573 was that ITIN mapping is deferred for some observer work. I don't know if that has changed.

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 6]
```diff
@@ -24,46 +67,166 @@ def generate_itin(
...
+        # Group numbers have range 50-65, 70-88, 90-92, or 94-99, inclusive
+        itin["group"] = random_integers(
+            min_val=1,
+            max_val=44,  # (65-50+1)+(88-70+1)+(92-90+1)+(99-94+1)
+            index=itin.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_itin_group",
+        )
+
+        # Rescale the group digits to match expected ranges
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:95]: clever

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py [hunk 7]
```diff
@@ -24,46 +67,166 @@ def generate_itin(
...
+
+        return itin["itin"].to_numpy()
+
+    if additional_key is None:
+        additional_key = 1
+    itins = pd.Series(generate_itin(size, additional_key))
+    duplicate_mask = itins.duplicated()
+    counter = 0
+    while duplicate_mask.sum() > 0:
+        if counter >= 10:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:125]: Is this limit a guess?

**Review by stevebachmeier — APPROVED** [review]

**Comment by aflaxman** [comment]
Superstition!  E.g. https://www.reuters.com/world/us/us-supreme-court-rebuffs-number-beast-religious-dispute-2021-06-28/

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py (hunk 3 above)
Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:169]: I forgot I left this comment! I think I had that comment as a question for RT, which I've got answered.

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py (hunk 4 above)
Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:194]: I'm using dot notation for the column access, which is also bad form.

Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:194]: I don't see a logical difference between `.astype(str)` and `.map(str)` but [this](https://stackoverflow.com/questions/50782934/what-is-difference-between-mapstr-and-astypestr-in-dataframe) suggests `astype` is more performant than `map` for this case.

Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:194]: I've updated this to be the one-line (pre-black) return for SSN and ITIN generation.

File: src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py (hunk 7 above)
Comment (mattkappel) [reply src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py:125]: yup!

**Pull request merged**